lua/starfall/libs_cl/xinput.lua

```diff
@@ -0,0 +1,140 @@
+	require("xinput")
+else
+	return
+end
+
+--- A simpler, hook-based, and more-powerful controller input library. Inputs are not lost between rendered frames, and there is support for rumble. Note: the client must have the XInput lua binary module installed in order to access this library. See more at https://github.com/mitterdoo/garrysmod-xinput
+-- @client
+local xinput_library = SF.RegisterLibrary("xinput")
+
+--- Gets the state of the controller.
+-- @name xinput_library.getState
+-- @class function
+-- @param id Controller number. Starts at 0
+-- @return Table containing all input data of the controller, or false if the controller is not connected. The table uses this struct: https://github.com/mitterdoo/garrysmod-xinput#xinput_gamepad
+xinput_library.getState = xinput.getState
+
+--- Gets whether the button on the controller is currently pushed down.
+-- @name xinput_library.getButton
+-- @class function
+-- @param id Controller number. Starts at 0
+-- @param button The button to check for. See https://github.com/mitterdoo/garrysmod-xinput#xinput_gamepad_
+-- @return bool
+xinput_library.getButton = xinput.getButton
+
+--- Gets the current position of the trigger on the controller.
+-- @name xinput_library.getTrigger
+-- @class function
+-- @param id Controller number. Starts at 0
+-- @param trigger Which trigger to use. 0 is left
+-- @return 0-255 inclusive
+xinput_library.getTrigger = xinput.getTrigger
+
+--- Gets the current coordinates of the stick on the controller.
+-- @name xinput_library.getStick
+-- @class function
+-- @param id Controller number. Starts at 0
+-- @param stick Which stick to use. 0 is left
+-- @return Two numbers for the X and Y coordinates, respectively, each being between -32768 - 32767 inclusive
+xinput_library.getStick = xinput.getStick
+
+--- Attempts to check the battery level of the controller.
+-- @name xinput_library.getBatteryLevel
+-- @class function
+-- @param id Controller number. Starts at 0
+-- @return If successful: a number between 0.0-1.0 inclusive. If unsuccessful: false, and a string error message
+xinput_library.getBatteryLevel = xinput.getBatteryLevel
+
+--- Gets all of the connected controllers.
+-- @name xinput_library.getControllers
+-- @class function
+-- @return A table where each key is the ID of the controller that is connected. Disconnected controllers are not placed in the table.
+xinput_library.getControllers = xinput.getControllers
+
+--- Sets the rumble on the controller.
+-- @param id Controller number. Starts at 0
+-- @param softPercent A number between 0.0-1.0 for how much the soft rumble motor should vibrate.
+-- @param hardPercent A number between 0.0-1.0 for how much the hard rumble motor should vibrate.
+function xinput_library.setRumble(id, softPercent, hardPercent)
+	-- This longer function makes sure that the rumble doesn't continue when the instance is gone.
+	SF.CheckLuaType(id, TYPE_NUMBER)
+	if id % 1 ~= 0 then
```

Comment:
The other functions don't throw when non-integer. Maybe just round down?

Reply:
The other functions convert it to an integer already. This one has to do it in Lua, so it can store the current rumble values in a table on the chip

Reply:
Oh I see what you mean. Yes I'll just do that